### PR TITLE
Fix Bridge pool recommendations

### LINKED ISSUE #1799: Fix Bridge pool recommendations SDK functionality
Related to #1704

As of #1795, the Bridge pool recommendations algorithm is "broken" (or more accurately, incomplete). It should take into account the fact that pending transfers in the Bridge pool might be paid in more gas token types other than NAM.

### PR DESCRIPTION
**Pull request by sug0**
## Describe your changes

Closes #1799

## Indicate on which release or other PRs this topic is based on

Based on #1795

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

**Review by batconjurer — CHANGES_REQUESTED** [review]
The code changes look fine, but needs to new unit tests to cover the new code paths for provided conversion tables.

**Comment by sug0** [comment]
@batconjurer  there's currently a gotcha in the recommendations algorithm, I think. the `transfer_to_ethereum_progress` RPC method skips "seen" Ethereum transfers. a transfer to be relayed is not recommended if it is not found in the array of transfers returned by `transfer_to_ethereum_progress`; in turn, all transfers considered for relaying are obtained from querying the current signed pool.

the current signed pool may be behind the most up to date Bridge pool, because of the vote extension tx lag. so, I believe the algorithm may recommend a transfer that has already been relayed, that still shows up in the signed pool.

a simple solution may be to return with an error if the signed Bridge pool nonce does not match the up to date Bridge pool nonce in storage. this way we never recommend transfers that have already been relayed and acted upon.

File: core/src/std_ext.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    fn maybe_collect_result<C>(self) -> Result<C, E>
+    where
+        C: Default + Extend<T>;
+}
+
+impl<I, T, E> MaybeCollectResult<T, E> for I
+where
+    I: Iterator<Item = Result<T, E>>,
+{
+    fn maybe_collect_result<C>(self) -> Result<C, E>
```
Comment (batconjurer) [inline core/src/std_ext.rs:16]: Is this used?

**Comment by sug0** [comment]
**TODO:** test recommending two different token types

File: core/src/std_ext.rs
Comment (tzemanovic) [inline core/src/std_ext.rs:1]: I think you can already do this via https://doc.rust-lang.org/stable/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA,+E%3E%3E-for-Result%3CV,+E%3E without `Default` and `Extend`. In bridge_pool.rs where it's used you can call `Result::from_iter` or assign the iter over result to `let iter_res: Result =`. But if you prefer the chained calls I don't mind keeping this, but I'd change it to:

```rust
pub trait MaybeCollectResult<T, E> {
    /// Collect all [`Ok`] values in the iterator, or fail.
    fn maybe_collect_result<C>(self) -> Result<C, E>
    where
        C: FromIterator<T>;
}

impl<I, T, E> MaybeCollectResult<T, E> for I
where
    I: Iterator<Item = Result<T, E>>,
{
    fn maybe_collect_result<C>(self) -> Result<C, E>
    where
        C: FromIterator<T>,
    {
        Result::from_iter(self)
    }
}
```

**Review by tzemanovic — APPROVED** [review]

**Pull request merged**